### PR TITLE
add reco/ID, resolution, scale uncertainties

### PR DESCRIPTION
**Pull request by JanFSchulte**
This PR adds the code for several mass dependent muon uncertainties in the histogrammer

### CODE REVIEW

File: ZprimeMuMuPatMiniAodNewMC.C [hunk 1]
```diff
@@ -1169,40 +1248,46 @@ void ZprimeMuMuPatMiniAodNewMC::PlotRecoInfo(float CosmicMuonRejec, float vertex
...
+  h2_CSMassMuIDBinned_   ->Fill(m_vtxMassMu,               (priEtaBin*3)+0,newweight*SF1*SF2);
+  h2_CSMassUpBinned_     ->Fill(m_vtxMassScaledMu,         (priEtaBin*3)+0,newweight);
+  h2_CSMassDownBinned_   ->Fill(m_vtxMassScaledMu          ,(priEtaBin*3)+0,newweight);
   if (secEtaBin > 0) {
     // std::cout << "secondary bin (" << secEtaBin << "*3)+0(" << (secEtaBin*3)+0 << ")" << std::endl;
     h2_CSSmearedMassBinned_->Fill(m_vtxMassSmearedMu,        (secEtaBin*3)+0,newweight);
     h2_CSMassBinned_       ->Fill(m_vtxMassMu,               (secEtaBin*3)+0,newweight);
-    h2_CSMassUpBinned_     ->Fill(m_vtxMassMu*(1+m_scaleUnc),(secEtaBin*3)+0,newweight);
-    h2_CSMassDownBinned_   ->Fill(m_vtxMassMu*(1-m_scaleUnc),(secEtaBin*3)+0,newweight);
+    h2_CSMassMuIDBinned_   ->Fill(m_vtxMassMu,               (secEtaBin*3)+0,newweight*SF1*SF2);
```
Comment (jsturdy) [inline ZprimeMuMuPatMiniAodNewMC.C:1258]: Is this (and elsewhere) a correction on the mass? And would it make more sense to do, e.g.,:
```
    h2_CSMassMuIDBinned_   ->Fill(m_vtxMassMu*SF1*SF2,               (secEtaBin*3)+0,newweight);
```
For another part of the code I was thinking about this in the past, and wasn't sure what the effect would be, and what the "correct" way was.

File: ZprimeMuMuPatMiniAodNewMC.C [hunk 2]
```diff
@@ -798,6 +808,61 @@ void ZprimeMuMuPatMiniAodNewMC::Loop(bool debug)
...
+     // Double smearing for muons in the endcaps
+     if (std::abs(eta) > 1.2) smearing *= 2;
+     // Return Gaussian smearing
+     double ratio = rand->Gaus(1, smearing);
+     return ratio;
+
+ 
+}
+
+TLorentzVector ZprimeMuMuPatMiniAodNewMC::GetShiftedMuon(float px, float py, float pz, float E, float ratio){
```
Comment (jsturdy) [inline ZprimeMuMuPatMiniAodNewMC.C:855]: Does this method leak (`new` not accompanied by `delete`)?
Could clean up with a `shared_ptr`

**Review by jsturdy — COMMENTED** [review]
Few questions

File: ZprimeMuMuPatMiniAodNewMC.C (hunk 2 above)
Comment (JanFSchulte) [reply ZprimeMuMuPatMiniAodNewMC.C:855]: Hmm, yeah, could be leaky. I'm not sure how these shared_ptr work, but if can point me to an example I can improve this.

File: ZprimeMuMuPatMiniAodNewMC.C (hunk 1 above)
Comment (JanFSchulte) [reply ZprimeMuMuPatMiniAodNewMC.C:1258]: No, SF1 and SF2 are per event weights as they represent an efficiency loss at high p. So I think multiplying it onto the weight is correct here.

Comment (jsturdy) [reply ZprimeMuMuPatMiniAodNewMC.C:1258]: OK, makes sense, thanks!

File: ZprimeMuMuPatMiniAodNewMC.C (hunk 2 above)
Comment (jsturdy) [reply ZprimeMuMuPatMiniAodNewMC.C:855]: I'd try something a la (I always have to trial a few things to get the syntax correct for the specific constructor):
```c++
std::shared_ptr<TLorentzVector> result = std::make_shared<TLorentzVector>()
result->SetPxPyPzE(ratio*px,
    ratio*py,
    ratio*pz,
    ratio*E);
```
or maybe
```c++
std::shared_ptr<TLorentzVector> result = std::make_shared<TLorentzVector>(ratio*px,ratio*py,ratio*pz,ratio*E);)
```

**Pull request merged**